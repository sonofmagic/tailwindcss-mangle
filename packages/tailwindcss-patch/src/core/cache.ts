import fs from 'node:fs'
import path from 'node:path'
import { pkgName } from '@/constants'
import type { CacheOptions, InternalCacheOptions } from '@/types'
import { log } from '@/logger'

export function getCacheOptions(options?: CacheOptions | boolean) {
  let cache: InternalCacheOptions
  switch (typeof options) {
    case 'undefined': {
      cache = {
        enable: false,
      }
      break
    }
    case 'boolean': {
      cache = {
        enable: options,
      }
      break
    }
    case 'object': {
      cache = { ...options, enable: true }
      break
    }
  }
  return cache
}

export class CacheManager {
  public options: Required<CacheOptions> & { filename: string }

  constructor(options: CacheOptions = {}) {
    this.options = this.getOptions(options)
  }

  mkdir(cacheDirectory: string) {
    const exists = fs.existsSync(cacheDirectory)
    if (!exists) {
      fs.mkdirSync(cacheDirectory, {
        recursive: true,
      })
    }
    return cacheDirectory
  }

  getOptions(options: CacheOptions = {}): Required<CacheOptions> & { filename: string } {
    const cwd = options.cwd ?? process.cwd()
    const dir = options.dir ?? path.resolve(cwd, 'node_modules/.cache', pkgName)
    const file = options.file ?? 'index.json'
    const filename = path.resolve(dir, file)
    return {
      cwd,
      dir,
      file,
      filename,
      strategy: 'merge',
    }
  }

  write(data: Set<string>) {
    try {
      const { dir, filename } = this.options
      this.mkdir(dir)
      fs.writeFileSync(filename, JSON.stringify([...data], undefined, 2), 'utf8')
      return filename
    }
    catch {
      log('write cache file fail!')
    }
  }

  read() {
    const { filename } = this.options
    try {
      if (fs.existsSync(filename)) {
        const data = fs.readFileSync(filename, 'utf8')
        return new Set<string>(JSON.parse(data))
      }
    }
    catch {
      log(`parse cache content fail! path:${filename}`)
      try {
        fs.unlinkSync(filename)
      }
      catch {
        log(`delete cache file fail! path:${filename}`)
      }
    }
  }
}
