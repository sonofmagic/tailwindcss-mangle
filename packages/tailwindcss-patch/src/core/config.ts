export * from '@tailwindcss-mangle/config'
