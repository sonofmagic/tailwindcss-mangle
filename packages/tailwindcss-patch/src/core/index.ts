export * from './tw-patcher'
export * from './exposeContext'
export * from './inspector'
export * from './runtime-patcher'
export * from './cache'
export * from './config'
