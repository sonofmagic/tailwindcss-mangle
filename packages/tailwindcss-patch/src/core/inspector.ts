import * as t from '@babel/types'
import { parse, traverse, generate } from '@/babel'
// crash code

export function inspectProcessTailwindFeaturesReturnContext(content: string) {
  const ast = parse(content)
  let hasPatched = false
  traverse(ast, {
    FunctionDeclaration(p) {
      const n = p.node
      if (n.id?.name === 'processTailwindFeatures' && n.body.body.length === 1 && t.isReturnStatement(n.body.body[0])) {
        const rts = n.body.body[0]
        if (t.isFunctionExpression(rts.argument)) {
          const body = rts.argument.body.body
          const lastStatement = body.at(-1)
          hasPatched = t.isReturnStatement(lastStatement) && t.isIdentifier(lastStatement.argument) && lastStatement.argument.name === 'context'
          if (!hasPatched) {
            // return context;
            const rts = t.returnStatement(t.identifier('context'))
            body.push(rts)
          }
        }
      }
    }
  })

  return {
    code: hasPatched ? content : generate(ast).code,
    hasPatched
  }
}

export function inspectPostcssPlugin(content: string) {
  const ast = parse(content)
  const exportKey = 'contextRef'
  const variableName = 'contextRef'
  const valueKey = 'value'
  let hasPatched = false
  traverse(ast, {
    Program(p) {
      const n = p.node
      // find module.exports = function tailwindcss(configOrPath)
      const idx = n.body.findIndex((x) => {
        return (
          t.isExpressionStatement(x) &&
          t.isAssignmentExpression(x.expression) &&
          t.isMemberExpression(x.expression.left) &&
          t.isFunctionExpression(x.expression.right) &&
          x.expression.right.id?.name === 'tailwindcss'
        )
      })

      if (idx > -1) {
        const prevStatement = n.body[idx - 1]
        const lastStatement = n.body.at(-1)
        const hasPatchedCondition0 =
          prevStatement &&
          t.isVariableDeclaration(prevStatement) &&
          prevStatement.declarations.length === 1 &&
          t.isIdentifier(prevStatement.declarations[0].id) &&
          prevStatement.declarations[0].id.name === variableName
        const hasPatchedCondition1 =
          t.isExpressionStatement(lastStatement) &&
          t.isAssignmentExpression(lastStatement.expression) &&
          t.isIdentifier(lastStatement.expression.right) &&
          lastStatement.expression.right.name === variableName

        hasPatched = hasPatchedCondition0 || hasPatchedCondition1
        if (!hasPatched) {
          // const contextRef = {
          //   value: []
          // };
          const statement = t.variableDeclaration('const', [
            t.variableDeclarator(t.identifier(variableName), t.objectExpression([t.objectProperty(t.identifier(valueKey), t.arrayExpression())]))
          ])
          n.body.splice(idx, 0, statement)
          // module.exports.contextRef = contextRef;
          n.body.push(
            t.expressionStatement(
              t.assignmentExpression(
                '=',
                t.memberExpression(t.memberExpression(t.identifier('module'), t.identifier('exports')), t.identifier(exportKey)),
                t.identifier(variableName)
              )
            )
          )
        }
      }
    },
    FunctionExpression(p) {
      if (hasPatched) {
        return
      }
      const n = p.node
      if (n.id?.name === 'tailwindcss' && n.body.body.length === 1 && t.isReturnStatement(n.body.body[0])) {
        const returnStatement = n.body.body[0]
        if (t.isObjectExpression(returnStatement.argument) && returnStatement.argument.properties.length === 2) {
          const properties = returnStatement.argument.properties
          if (t.isObjectProperty(properties[0]) && t.isObjectProperty(properties[1])) {
            const keyMatched = t.isIdentifier(properties[0].key) && properties[0].key.name === 'postcssPlugin'
            const pluginsMatched = t.isIdentifier(properties[1].key) && properties[1].key.name === 'plugins'
            if (
              pluginsMatched &&
              keyMatched &&
              t.isCallExpression(properties[1].value) &&
              t.isMemberExpression(properties[1].value.callee) &&
              t.isArrayExpression(properties[1].value.callee.object)
            ) {
              const pluginsCode = properties[1].value.callee.object.elements
              if (pluginsCode[1] && t.isFunctionExpression(pluginsCode[1])) {
                const targetBlockStatement = pluginsCode[1].body

                const lastStatement = targetBlockStatement.body.at(-1)
                if (t.isExpressionStatement(lastStatement)) {
                  // contextRef.value.push((0, _processTailwindFeatures.default)(context)(root, result));
                  const newExpressionStatement = t.expressionStatement(
                    t.callExpression(
                      t.memberExpression(
                        t.memberExpression(t.identifier(variableName), t.identifier('value')),

                        t.identifier('push')
                      ),

                      [lastStatement.expression]
                    )
                  )
                  targetBlockStatement.body[targetBlockStatement.body.length - 1] = newExpressionStatement
                }

                const ifIdx = targetBlockStatement.body.findIndex((x) => t.isIfStatement(x))
                if (ifIdx > -1) {
                  const ifRoot = <t.IfStatement>targetBlockStatement.body[ifIdx]
                  if (t.isBlockStatement(ifRoot.consequent) && ifRoot.consequent.body[1] && t.isForOfStatement(ifRoot.consequent.body[1])) {
                    const forOf: t.ForOfStatement = ifRoot.consequent.body[1]
                    if (t.isBlockStatement(forOf.body) && forOf.body.body.length === 1 && t.isIfStatement(forOf.body.body[0])) {
                      const if2: t.IfStatement = forOf.body.body[0]
                      if (t.isBlockStatement(if2.consequent) && if2.consequent.body.length === 1 && t.isExpressionStatement(if2.consequent.body[0])) {
                        const target = if2.consequent.body[0]
                        // contextRef.value.push((0, _processTailwindFeatures.default)(context)(root1, result));
                        const newExpressionStatement = t.expressionStatement(
                          t.callExpression(t.memberExpression(t.memberExpression(t.identifier(variableName), t.identifier('value')), t.identifier('push')), [target.expression])
                        )
                        if2.consequent.body[0] = newExpressionStatement
                      }
                    }
                  }
                }
                // clear contextRef.value
                targetBlockStatement.body.unshift(
                  // contentRef.value = []
                  // t.expressionStatement(t.assignmentExpression('=', t.memberExpression(t.identifier(variableName), t.identifier(valueKey)), t.arrayExpression()))

                  // contentRef.value.length = 0
                  t.expressionStatement(
                    t.assignmentExpression(
                      '=',
                      t.memberExpression(t.memberExpression(t.identifier(variableName), t.identifier(valueKey)), t.identifier('length')),
                      t.numericLiteral(0)
                    )
                  )
                )
              }
            }
          }
        }
      }
      // start = true
    }
    // BlockStatement(p) {
    //   const n = p.node
    //   if (start && p.parent.type === 'FunctionExpression' && !p.parent.id) {
    //     n.body.unshift(t.expressionStatement(t.assignmentExpression('=', t.memberExpression(t.identifier(variableName), t.identifier(valueKey)), t.arrayExpression())))
    //   }
    // }
  })
  return {
    code: hasPatched ? content : generate(ast).code,
    hasPatched
  }
}
