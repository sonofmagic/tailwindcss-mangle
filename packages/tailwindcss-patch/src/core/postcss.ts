import postcss from 'postcss'
import path from 'node:path'

export async function processTailwindcss(options: { cwd?: string; config?: string }) {
  options.cwd = options.cwd ?? process.cwd()
  const config = options.config && path.isAbsolute(options.config) ? options.config : path.resolve(options.cwd, 'tailwind.config.js')

  const res = await postcss([
    require('tailwindcss')({
      config
    })
  ]).process('@tailwind base;@tailwind components;@tailwind utilities;', {
    from: undefined
  })
  return res
}
