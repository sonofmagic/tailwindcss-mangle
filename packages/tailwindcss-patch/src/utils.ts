import fss from 'node:fs'
import fs from 'node:fs/promises'
import type { SyncOpts } from 'resolve'
import pkg from 'resolve'

const { sync } = pkg

export function ensureFileContent(filepaths: string | string[]) {
  if (typeof filepaths === 'string') {
    filepaths = [filepaths]
  }
  let content
  for (const filepath of filepaths) {
    if (fss.existsSync(filepath)) {
      content = fss.readFileSync(filepath, {
        encoding: 'utf8',
      })
      break
    }
  }
  return content
}

export function requireResolve(id: string, opts?: SyncOpts) {
  return sync(id, opts)
}

export async function ensureDir(p: string) {
  try {
    await fs.access(p)
  }
  catch {
    await fs.mkdir(p, {
      recursive: true,
    })
  }
}
