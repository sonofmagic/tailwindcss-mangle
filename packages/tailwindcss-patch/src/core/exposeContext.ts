import path from 'node:path'
import fs from 'node:fs'
import type { TailwindcssClassCache, TailwindcssRuntimeContext } from '@/types'
import { requireResolve } from '@/utils'

export function getTailwindcssEntry(basedir?: string) {
  return requireResolve('tailwindcss', {
    basedir,
  })
}

export function getContexts(basedir?: string): TailwindcssRuntimeContext[] {
  const twPath = getTailwindcssEntry(basedir)

  const distPath = path.dirname(twPath)

  let injectFilePath = path.join(distPath, 'plugin.js')
  if (!fs.existsSync(injectFilePath)) {
    injectFilePath = path.join(distPath, 'index.js')
  }

  const mo = require(injectFilePath)
  if (mo.contextRef) {
    return mo.contextRef.value
  }
  return []
}

export function getClassCaches(basedir?: string): TailwindcssClassCache[] {
  const contexts = getContexts(basedir)
  return contexts.map(x => x.classCache)
}

export function getClassCacheSet(basedir?: string, options?: { removeUniversalSelector?: boolean }): Set<string> {
  const classCaches = getClassCaches(basedir)
  const classSet = new Set<string>()
  for (const classCacheMap of classCaches) {
    const keys = classCacheMap.keys()
    for (const key of keys) {
      const v = key.toString()
      if (options?.removeUniversalSelector && v === '*') {
        continue
      }
      classSet.add(v)
    }
  }
  return classSet
}
