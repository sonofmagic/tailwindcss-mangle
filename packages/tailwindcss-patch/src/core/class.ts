import { getClassCacheSet, getContexts, getTailwindcssEntry } from './exposeContext'
import type { CacheOptions, InternalCacheOptions, InternalPatchOptions, TailwindcssPatcherOptions, CacheStrategy } from '../types'
import { writeCache, readCache } from './cache'
import { createPatch, getPatchOptions } from './patcher'

export function getCacheOptions(options?: CacheOptions | boolean) {
  let cache: InternalCacheOptions
  switch (typeof options) {
    case 'undefined': {
      cache = {
        enable: false
      }
      break
    }
    case 'boolean': {
      cache = {
        enable: options
      }
      break
    }
    case 'object': {
      cache = { ...options, enable: true }
      break
    }
  }
  return cache
}
export class TailwindcssPatcher {
  public rawOptions: TailwindcssPatcherOptions
  public cacheOptions: InternalCacheOptions
  public patchOptions: InternalPatchOptions
  public patch: () => void
  constructor(options: TailwindcssPatcherOptions = {}) {
    this.rawOptions = options
    this.cacheOptions = getCacheOptions(options.cache)
    this.patchOptions = getPatchOptions(options.patch)
    this.patch = createPatch(this.patchOptions)
  }

  getPkgEntry(basedir?: string) {
    return getTailwindcssEntry(basedir)
  }

  setCache(set: Set<string>) {
    if (this.cacheOptions.enable) {
      return writeCache(set, this.cacheOptions)
    }
  }

  getCache() {
    // if(this.cache.enable){
    return readCache(this.cacheOptions)
    // }
  }

  /**
   * @description 在多个 tailwindcss 上下文时，这个方法将被执行多次，所以策略上应该使用 append
   * 详见 taro weapp-tailwindcss 独立分包
   * @param basedir
   * @returns
   */
  getClassSet(
    options: {
      basedir?: string
      cacheStrategy?: CacheStrategy
    } = {
      cacheStrategy: this.cacheOptions.strategy ?? 'merge'
    }
  ) {
    const { basedir, cacheStrategy } = options
    const set = getClassCacheSet(basedir)
    if (cacheStrategy === 'overwrite') {
      set.size > 0 && this.setCache(set)
    } else if (cacheStrategy === 'merge') {
      const cacheSet = this.getCache()
      if (cacheSet) {
        for (const x of cacheSet) {
          set.add(x)
        }
      }
      this.setCache(set)
    }

    return set
  }

  getContexts(basedir?: string) {
    return getContexts(basedir)
  }
}
