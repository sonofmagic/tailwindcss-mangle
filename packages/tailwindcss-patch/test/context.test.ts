import { getCss, getTestCase } from './utils'
import { getClassCacheSet, getContexts } from '@/core'

describe('common usage', () => {
  it('hello-world', async () => {
    const result = await getCss(getTestCase('hello-world.html'))
    const ctxs = getContexts()
    expect(ctxs.length).toBeTruthy()
    const set = getClassCacheSet()
    expect(set.size).toBeGreaterThan(0)
    expect(set.size).toBe(4)
    expect(result).toMatchSnapshot()
  })

  it('hello-world with js', async () => {
    const result = await getCss([getTestCase('hello-world.html'), getTestCase('hello-world.js')])
    const ctxs = getContexts()
    expect(ctxs.length).toBeTruthy()
    const set = getClassCacheSet()
    expect(set.size).toBeGreaterThan(0)
    expect(set.size).toBe(5)
    expect(result).toMatchSnapshot()
  })

  // https://github.com/sonofmagic/weapp-tailwindcss-webpack-plugin/issues/158
  it('bg-[url(\'img_src\')] lose efficacy', async () => {
    await getCss([getTestCase('img-url.jsx')])
    const ctxs = getContexts()
    expect(ctxs).toBeTruthy()
    const set = getClassCacheSet()
    expect(set.size).toBeGreaterThan(0)
    expect(set.size).toBe(2)
    expect([...set.values()][1]).toBe('bg-[url(\'https://xxx.webp\')]')
    //
  })

  it('trailing slash', async () => {
    await getCss([getTestCase('trailing-slash.vue')])
    const ctxs = getContexts()
    expect(ctxs).toBeTruthy()
    const set = getClassCacheSet()
    expect(set.size).toBeGreaterThan(0)
    expect(set.size).toBe(4)
    expect(set.has('bg-red-500')).toBe(true)
    expect(set.has('bg-red-500/50')).toBe(true)
  })
})
