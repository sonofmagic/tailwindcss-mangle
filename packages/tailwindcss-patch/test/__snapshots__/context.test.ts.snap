// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`common usage > hello-world 1`] = `
".text-3xl {
    font-size: 1.875rem;
    line-height: 2.25rem
}
.font-bold {
    font-weight: 700
}
.underline {
    text-decoration-line: underline
}"
`;

exports[`common usage > hello-world with js 1`] = `
".bg-\\\\[\\\\#123456\\\\] {
    --tw-bg-opacity: 1;
    background-color: rgb(18 52 86 / var(--tw-bg-opacity))
}
.text-3xl {
    font-size: 1.875rem;
    line-height: 2.25rem
}
.font-bold {
    font-weight: 700
}
.underline {
    text-decoration-line: underline
}"
`;
